src/swarm/neo/util/StructSerializer.d

```diff
@@ -0,0 +1,347 @@
+    enableStomping(buf);
+    auto serialized_struct = cast(S*)buf.ptr;
+    *serialized_struct = s;
+
+    // Handle any dynamic array fields.
+    seriaizeStructDynamicArrays(*serialized_struct, buf);
+}
+
+/*******************************************************************************
+
+    Interprets the data in the provided buffer (which must have been created by
+    a call to `serialize`, above) as a struct of the specific type.
+
+    Params:
+        S = type to deserialize
+        serialized = buffer containing the serialized thing
+
+    Returns:
+        pointer to the buffer interpreted as a pointer to an instance of the
+        serialized type
+
+*******************************************************************************/
+
+public S* deserialize ( S ) ( Const!(void)[] serialized )
+{
+    assert(serialized.length >= S.sizeof);
+    return cast(S*)serialized.ptr;
+}
+
+/*******************************************************************************
+
+    Serializes dynamic arrays inside a struct into the provided buffer. Both
+    dynamic array fields directly contained in the struct and dynamic array
+    fields of nested structs are handled. The contents of any dynamic arrays
+    found will be serialized into `buf` and their slices updated to point to the
+    content just added to the serialization buffer.
+
+    Params:
+        S = type of struct to serialize dynamic arrays from
+        s = struct instance to serialize dynamic arrays from
+        buf = buffer to serialize into
+
+*******************************************************************************/
+
+private void seriaizeStructDynamicArrays ( S ) ( ref S s, ref void[] buf )
+{
+    foreach ( ref f; s.tupleof )
+    {
+        // Serialize dynamic array fields of S.
+        static if ( isDynamicArrayType!(typeof(f)) )
+            serializeDynamicArray(f, buf);
+        // Recursively serialize dynamic arrays in struct fields of S.
+        else static if ( ContainsDynamicArray!(typeof(f)) )
+            seriaizeStructDynamicArrays(f, buf);
+    }
+}
+
+/*******************************************************************************
+
+    Serializes an array into the provided buffer. This has three steps:
+        1. The array's content is appended to the serialization buffer.
+        2. The array slice is updated to point to the content just added to the
+           serialization buffer.
+        3. If the array's elements are, in turn, dynamic arrays, repeats the
+           process (recursively) for each element.
+
+    Params:
+        A = type of array to serialize
+        array = slice to serialize and update to point to the serialized content
+        buf = buffer to serialize into
+
+*******************************************************************************/
+
+private void serializeDynamicArray ( A ) ( ref A array, ref void[] buf )
+{
+    static assert(isDynamicArrayType!(A));
+    static assert(isPrimitiveType!(BaseTypeOfArrays!(A)));
```

Comment:
Suggest
- using `A` as the array *element* type (unless is causes some type qualifier related complication) and
- disallowing indirections other than arrays,  i.e. also allowing struct/union/static array elements as long as they are plain values, instead of allowing elements of primitive type only.
```d
private void serializeDynamicArray ( A ) ( ref A[] array, ref void[] buf )
{
    // ...

    // Recursively serialize arrays of arrays.
    static if (isDynamicArrayType!(A))
    {
        foreach ( ref e; array )
            serializeDynamicArray(e, buf);
    }
    else
        static assert(!hasIndirections!(A));
}
```